File: app/src/main/java/ltd/kaizo/mynews/Controller/Fragments/BaseFragment.java
package ltd.kaizo.mynews.Controller.Fragments;

import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import icepick.Icepick;

public abstract class BaseFragment extends Fragment {

//    protected abstract BaseFragment newInstance(int position);

    protected abstract int getFragmentLayout();

    protected abstract void configureDesign();

    protected abstract void updateDesign();

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(getFragmentLayout(), container, false);
        ButterKnife.bind(this, view);
        this.configureDesign();
        return view;
    }

    @Override

    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        Icepick.restoreInstanceState(this, savedInstanceState);
        this.updateDesign();

    }


    @Override

    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        Icepick.saveInstanceState(this, outState);

    }
}

File: app/src/main/java/ltd/kaizo/mynews/controller/ui/base/BaseFragment.kt
package ltd.kaizo.mynews.controller.ui.base

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import butterknife.ButterKnife

/**
 * The type Base fragment.
 */
abstract class BaseFragment : Fragment() {

    /**
     * Gets fragment layout.
     *
     * @return the fragment layout
     */
    protected abstract val fragmentLayout: Int

    /**
     * Configure design.
     */
    protected abstract fun configureDesign()

    /**
     * Update design.
     */
    protected abstract fun updateDesign()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(fragmentLayout, container, false)
        this.configureDesign()
        return view
    }
    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        this.configureViewModel()
        this.configureObserver()
        this.updateDesign()
    }
    abstract fun configureViewModel()
    abstract fun configureObserver()

}
